### PR TITLE
Basic server-side implementation of configurable expectation proposal

### PR DESCRIPTION
**Pull request by GabrielKS**
What this PR _does_:
  * Provides a sample expectation/notification configuration file and a complete schema for such files
  * Exposes functions to calculate expectation/notification settings for a given trip based on its end date and inferred labels
  * Adds a stage to the intake pipeline that adds expectation data to each trip
  * Sends expectation data to the client app (which currently ignores it)
  * Fixes a bug in `TestLabelInferencePipeline.py`
  * Includes unit tests for most of the above

What this PR _doesn't do_:
  * Validate the expectation/notification configuration file against the schema
  * Fully implement the expectation proposal — for now, the only `expect` types implemented are `"all"` and `"none"`; the others default to `"all"`
  * Implement any notification behavior aside from the logic to calculate what the notifications should be

### CODE REVIEW

**Comment by GabrielKS** [comment]
The everlasting [UI proposal](https://github.com/e-mission/e-mission-docs/files/6743130/UI.Draft.2.pdf), for reference — though it's a bit out of date and could use a new draft. A better reference for the config file is the schema itself.

File: emission/tests/pipelineTests/TestExpectationPipeline.py [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+        primary_algorithms_stash = eacilp.primary_algorithms
+        eacilp.primary_algorithms = algorithms
+        test_options_stash = eaue._test_options
+        eaue._test_options["preprocess_trip"] = lambda trip: self.preprocess(trip)
+        etc.runIntakePipeline(self.testUUID)
+        eacilp.primary_algorithms = primary_algorithms_stash
+        eaue._test_options = test_options_stash
+    
+    def preprocess(self, trip):
+        # See eacilp.placeholder_predictor_2 for an explanation of the "fingerprint" technique
```
Comment (shankari) [inline emission/tests/pipelineTests/TestExpectationPipeline.py:52]: I don't see this explanation here - only a one-liner about `Timestamp2index`.

File: emission/analysis/userinput/expectations.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    expected_trip["data"]["inferred_trip"] = inferred_trip.get_id()
+    expected_trip["data"]["expectation"] = expectation
+    ts.insert(expected_trip)
+
+# This is a placeholder. TODO: implement the real algorithm
+def _get_expectation_for_trip(trip):
+    raw_expectation = eace.get_expectation(trip)
+    # For now, expect always labeling unless the config file specifies no labeling at all
+    processed_expectation = not raw_expectation["type"] == "none"
+    return {"to_label": processed_expectation}
```
Comment (shankari) [inline emission/analysis/userinput/expectations.py:52]: Can you add in a comment about what the difference between `raw` and `processed` is and is intended to be?
My understanding is that the raw expectation is of the form

```
                    "expect": {
                        "type": "randomDays",
                        "value": 2
                    },
```

and the processed is a simple boolean true/false on whether trips need to be labeled or not.

File: emission/tests/pipelineTests/TestExpectationPipeline.py [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+            494: None,
+            565: True,
+            795: False,
+            805: True,
+            880: True,
+            960: None
+        }
+        for trip in self.expected_trips:
+            fingerprint = trip["data"]["start_local_dt"]["hour"]*60+trip["data"]["start_local_dt"]["minute"]
+            if answers[fingerprint] is not None: self.assertEqual(trip["data"]["expectation"]["to_label"], answers[fingerprint])
```
Comment (shankari) [inline emission/tests/pipelineTests/TestExpectationPipeline.py:96]: why do you need this? won't the `"randomDays"` expectations also return `True` with your current implementation?

File: emission/tests/pipelineTests/TestExpectationNotificationConfig.py [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+                for i_rule,rule in enumerate(mode["rules"]):
+                    rule = mode["rules"][i_rule]
+                    trip = {"data": {"inferred_labels": labels}}
+                    guess = eace._rule_matches_label(mode, rule, trip, "label1")
+                    if i_rule in answer: self.assertTrue(guess)
+                    else: self.assertFalse(guess)
+
+
+    def testGetRuleForLabel(self):
+        # An answers entry is a (labels, index of corresponding rule under intensive mode, index of corresponding rule under relaxed mode) tuple
```
Comment (shankari) [inline emission/tests/pipelineTests/TestExpectationNotificationConfig.py:137]: nit: expand on this a bit to explain how you go from the list of all rules to the selected rule.

**Review by shankari — CHANGES_REQUESTED** [review]
I don't see any code changes required; these are primarily requests for clarification/additional documentation to improve long-term maintainability.

File: conf/ux/expectations.conf.schema.json [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+            "description": "The various modes we can be in and their specifications. The active mode is the first mode in this list with a schedule matching the current date or with no schedule.",
+            "type": "array",
+            "items": {
+                "$ref": "#/definitions/modeType"
+            },
+            "contains": {
+                "$comment": "Must contain a \"default\" element -- one with no schedule",
+                "not": {"required": ["schedule"]}
+            }
+        }
```
[the remark this answers is not in the document]
Comment (GabrielKS) [reply conf/ux/expectations.conf.schema.json:17]: Exactly. The scheduling algorithm iterates through modes from top to bottom and applies the first mode that either has a schedule matching the current date or has no schedule. The schema requires a mode with no schedule so that some mode will always apply.

File: conf/ux/expectations.conf.schema.json [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+    "required": ["modes"],
+    "definitions": {
+        "modeType": {
+            "description": "Specifies a collection mode that defines expectation and notification settings for a given time period",
+            "type": "object",
+            "properties": {
+                "enabled": {
+                    "description": "Whether or not to enable this mode; mode is enabled if this property is omitted",
+                    "type": "boolean"
+                },
```
[the remark this answers is not in the document]
Comment (GabrielKS) [reply conf/ux/expectations.conf.schema.json:28]: I just thought I'd add a way to "comment out" a mode. It's not strictly necessary. Do you think it adds excessive complexity?

File: conf/ux/expectations.conf.schema.json [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+                "label": {
+                    "description": "A human-readable label for the mode",
+                    "type": "string"
+                },
+                "confidenceThreshold": {
+                    "description": "Only display yellow labels with confidence greater than this threshold (0 to display all, 1 to display none)",
```
[the remark this answers is not in the document]
Comment (GabrielKS) [reply conf/ux/expectations.conf.schema.json:34]: This is actually something I haven't figured out yet. If the confidence for a given label is below the threshold, the phone should always display that label as red and blank; that part I'm sure of (and have implemented accordingly on the client side). The current implementation has these below-threshold labels also counting as red for the purposes of rule matching as well, but I'm not sure what the best choice is there.

If we wanted to be maximally flexible, we would _not_ have below-threshold labels match `"trigger": -1` because there is a semantic difference between very low confidence inferences and no inferences at all, and if a study admin wanted to treat them the same they could just write one more rule to explicitly match these trips. The argument in favor of having below-threshold labels match `"trigger": -1` (the status quo) is that it forces expectation/notification behavior to match what the user actually sees: to the user, red labels are red labels, if some trips that display red labels are appearing in `To Label` and some aren't (and it's not because of some random sample option), they might get confused. It's also simpler.

File: conf/ux/expectations.conf.schema.json [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+                }
+            },
+            "required": ["label", "confidenceThreshold", "rules"]
+        },
+        "scheduleType": {
+            "description": "Describes when the mode is active; omit to make always active",
+            "type": "object",
+            "properties": {
+                "startDate": {
+                    "description": "The base date from which to measure time; omit time zone to signify user's local time",
```
[the remark this answers is not in the document]
Comment (GabrielKS) [reply conf/ux/expectations.conf.schema.json:71]: The logic for this part is actually relatively simple; the phone's time zone is embedded in the trip data object, so the server just inserts it if we're missing a time zone. (The complicated part is the schedule matching, regardless of time zones, and in particular figuring out how many months/weeks/days there are between two dates — see `expectation_notification_config.py:64`).

Let's say we want to do primary data collection for one week, Monday to Sunday, every eight weeks. For simplicity, say we want the user to label all trips at the end of the day under primary data collection and none under secondary data collection. Under the existing schema, we can implement this as a schedule that begins at `00:00` user's local time and lasts `7` days, and specify a `startDate` that is a Monday. Then the user will have trips appear on their `To Label` screen for exactly 7 days every 56 days; if they label trips at the end of each day as they are supposed to, they will be interacting with the app 1 in 8 days.

Say we remove the ability to specify a time in the user's local time zone. We implement the schedule as beginning at `00:00` UTC. If the user is in a US time zone, primary mode will begin on Sunday evening and last until the next Sunday evening. Though this is the same length of time and on average will cover the same number of trips, it now spans 8 days instead of 7. If the user is labeling at the end of each day, they will be interacting with the app 1 in 7 days. This is worse.

To summarize, the ability to use a user's local time zone minimizes the number of days the user has to interact with the app. Even if they'll be labeling the same number of trips either way, every day we can avoid making the user do something is a hassle avoided.

I suppose if we wanted to simplify we could _only_ use local time and not give the option to specify a time zone.

File: emission/analysis/configs/expectation_notification_config.py [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+_test_options = {
+    "use_sample": False,
+    "override_keylist": None
+}
+
+# PRIVATE HELPER FUNCTIONS
+def _get_expectation_config():
+    if _test_options["use_sample"]:
+        config_file = open('conf/ux/expectations.conf.json.sample')
+    else:
```
[the remark this answers is not in the document]
Comment (GabrielKS) [reply emission/analysis/configs/expectation_notification_config.py:18]: As-is, no, but I want the tests to still work if someone adds a non-sample config file.

File: emission/tests/pipelineTests/TestExpectationNotificationConfig.py [hunk 10]
```diff
@@ -0,0 +1,206 @@
...
+    def testLoading(self):
+        mode_labels = [mode["label"] for mode in eace._config["modes"]]
+        self.assertEqual(len(mode_labels), 2)
+        self.assertEqual(set(mode_labels), {"intensive", "relaxed"})
+
+
+    def testParseDatetimeMaybeNaive(self):
+        # Etc/GMT+6 really means UTC-06:00 :(
+        self.assertEqual(eace._parse_datetime_maybe_naive("2021-07-09T12:34:56.789", "Etc/GMT+6"), arrow.get("2021-07-09T12:34:56.789-06:00"))
+        self.assertEqual(eace._parse_datetime_maybe_naive("2021-07-09T12:34:56.789+02:00", "Etc/GMT+6"), arrow.get("2021-07-09T12:34:56.789+02:00"))
```
[the remark this answers is not in the document]
Comment (GabrielKS) [reply emission/tests/pipelineTests/TestExpectationNotificationConfig.py:57]: So I don't have to think about daylight savings time. Arrow can parse all standard time zone strings; `America/Chicago` will not be a problem in practice.

File: emission/core/wrapper/expectedtrip.py [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+import emission.core.wrapper.trip as ecwt
+import emission.core.wrapper.wrapperbase as ecwb
+
+class Expectedtrip(ecwt.Trip):
+    props = ecwt.Trip.props
+    props.update({"raw_trip": ecwb.WrapperBase.Access.WORM,
+                  "cleaned_trip": ecwb.WrapperBase.Access.WORM,
+                  "inferred_labels": ecwb.WrapperBase.Access.WORM,
+                  "inferred_trip": ecwb.WrapperBase.Access.WORM,
+                  "expectation": ecwb.WrapperBase.Access.WORM
```
[the remark this answers is not in the document]
Comment (GabrielKS) [reply emission/core/wrapper/expectedtrip.py:17]: Will do in an upcoming commit.

File: emission/analysis/configs/expectation_notification_config.py [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+        if _rule_matches_label(mode, rule, trip, label): rules.append(rule)
+    return None if len(rules) == 0 else min(rules, key = lambda r: r["trigger"])
+
+def _get_rule(trip):
+    if len(_keylist) == 0: return _get_done_rule()
+    mode = _get_collection_mode_by_schedule(trip)
+    rules = []
+    for input in _keylist:
+        rule = _get_rule_for_label(mode, trip, input)
+        if rule is not None: rules.append(rule)
```
[the remark this answers is not in the document]
Comment (GabrielKS) [reply emission/analysis/configs/expectation_notification_config.py:108]: Yes, it is there for maximum flexibility. In `placeholder_predictor_2` I include an inference that only has one of the labels (`emission.analysis.classification.inference.labels.pipeline.py:71`). I never made a final decision on whether that should be allowed or not, so I allow it here and then throw an error a few lines later if it ends up not giving us any rules to work with.

[the remark this answers is not in the document]
Comment (GabrielKS) [reply emission/analysis/configs/expectation_notification_config.py:108]: Should I document or remove?

File: emission/analysis/userinput/expectations.py [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+    if _test_options["preprocess_trip"] is not None: _test_options["preprocess_trip"](expected_trip)
+
+    expectation = _get_expectation_for_trip(expected_trip)
+    # For now, I don't think it's necessary to save each expectation as its own database entry
+
+    expected_trip["data"]["inferred_trip"] = inferred_trip.get_id()
+    expected_trip["data"]["expectation"] = expectation
+    ts.insert(expected_trip)
+
+# This is a placeholder. TODO: implement the real algorithm
```
[the remark this answers is not in the document]
Comment (GabrielKS) [reply emission/analysis/userinput/expectations.py:47]: Currently it only tests whether `expect.type` is `"none"` or something else. It works for `"none"` vs. `"all"`, but eventually I'd like to implement the random sample stuff described in the schema.

File: emission/tests/pipelineTests/TestExpectationPipeline.py (hunk 1 above)
Comment (GabrielKS) [reply emission/tests/pipelineTests/TestExpectationPipeline.py:52]: I can flesh out that explanation in an upcoming commit.

File: emission/analysis/userinput/expectations.py (hunk 2 above)
Comment (GabrielKS) [reply emission/analysis/userinput/expectations.py:52]: That's correct. Will add the documentation in an upcoming commit.

File: emission/tests/pipelineTests/TestExpectationPipeline.py (hunk 3 above)
Comment (GabrielKS) [reply emission/tests/pipelineTests/TestExpectationPipeline.py:96]: I want the tests to enforce the actual intended behavior, not the current implementation. The actual intended behavior for the random options is not currently well-defined, so we don't test it — returning `True` for the random options is an implementation detail (and a placeholder, at that).

File: conf/ux/expectations.conf.schema.json (hunk 6 above)
[the remark this answers is not in the document]
Comment (shankari) [reply conf/ux/expectations.conf.schema.json:28]: no, that's fine. Just update the documentation in a future commit to highlight this.

File: conf/ux/expectations.conf.schema.json (hunk 7 above)
[the remark this answers is not in the document]
Comment (shankari) [reply conf/ux/expectations.conf.schema.json:34]: I agree that this is complex. Let's discuss it after tomorrow. I really want to deploy this today so we can walk through the new screen at tomorrow's deployer/beta tester meeting.

File: conf/ux/expectations.conf.schema.json (hunk 8 above)
[the remark this answers is not in the document]
Comment (shankari) [reply conf/ux/expectations.conf.schema.json:71]: Again, let's discuss this after tomorrow. Right now, I'm going to merge this and get the actual recommendations to show up.

File: emission/tests/pipelineTests/TestExpectationNotificationConfig.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (shankari) [reply emission/tests/pipelineTests/TestExpectationNotificationConfig.py:57]: ah, makes sense.

File: emission/analysis/configs/expectation_notification_config.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (shankari) [reply emission/analysis/configs/expectation_notification_config.py:108]: Document. We encountered one example already (Ft. Collins) where users did not fill in the replaced mode.

**Review by shankari — APPROVED** [review]
I am going to go ahead and merge this now so we can deploy tonight, demo tomorrow and start getting feedback. I've left the "will handle in subsequent commit" comments unresolved for your reference.

File: emission/tests/pipelineTests/TestExpectationNotificationConfig.py (hunk 4 above)
Comment (shankari) [reply emission/tests/pipelineTests/TestExpectationNotificationConfig.py:137]: Please handle this in a subsequent commit.

Comment (GabrielKS) [reply emission/tests/pipelineTests/TestExpectationNotificationConfig.py:137]: I'm not sure I understand what you mean.
